### PR TITLE
Split yang tests into multiple files

### PR DESCRIPTION
**Pull request by joyas-joseph**
Signed-off-by: Praveen Chaudhary <pchaudhary@linkedin.com>
Signed-off-by: Joyas Joseph <joyas_joseph@dell.com>

Link to previous PR  [#3730](https://github.com/Azure/sonic-buildimage/pull/3730)

#### Why I did it
Add manageability to the yang model tests 

#### How I did it
By splitting the tests and config data for the tests into multiple files.
The "tests" directory contains all the tests and the "tests_config" directory contains the configs used for the tests.

New tests can be added in new json files.

#### How to verify it
Build the sonic_yang_models package

#### Which release branch to backport (provide reason below if selected)
- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012

#### Description for the changelog
Improve yang model test manageability

### CODE REVIEW

**Comment by lguohan** [comment]
can you check the build failure?

**Comment by joyas-joseph** [comment]
> can you check the build failure?

I have updated the PR. All checks pass now.

**Comment by joyas-joseph** [comment]
@praveen-li Could you please review this?

**Comment by praveen-li** [comment]
> @praveen-li Could you please review this?

sure I'll have a look today thanks

File: src/sonic-yang-models/tests/yang_model_tests/tests/acl.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests/acl.json:4]: If Libyang changes the Error string in the next version, We may have to change it at multiple places. I think, we should think a better way to write errors, so that we change only at few places later.

File: src/sonic-yang-models/tests/yang_model_tests/tests/acl.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests/acl.json:8]: Instead, we can say None, and change the test code to Handle None as []

File: src/sonic-yang-models/tests/yang_model_tests/test_yang_model.py
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/test_yang_model.py:123]: We have already loaded the entire config before. So using ijson now will not have much benefit.

File: src/sonic-yang-models/tests/yang_model_tests/test_yang_model.py
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/test_yang_model.py:57]: This json load should be fine, since test json files should not grow super large.  But lets see, if we can find better way, so we do not load entire test config files in a single shot. I think we can have a section for test_config_file in tests/*.json, and while parsing here, we can create self.testFile = {} as well.

File: src/sonic-yang-models/tests/yang_model_tests/tests/crm.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests/crm.json:2]: Say we have one more field
{
'test_config_file': 'crm.json'
}
This will be parsed as non test and will help to create self.testFile dict.

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json:1]: Should be way something like sonic_specific_table.json

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json:58]: Shd go in CRM or in FLEX counter.

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json:233]: VLAN.json

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json:250]: flex_counter.json

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json:283]: vlan.json

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json:300]: interface.json

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json:425]: port_channel.json

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json:470]: port.json

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/except.json:622]: I feel, this file still has mixed config.

**Review by praveen-li — COMMENTED** [review]
Thanks for starting this effort.

File: src/sonic-yang-models/tests/yang_model_tests/test_yang_model.py
Comment (joyas-joseph) [reply src/sonic-yang-models/tests/yang_model_tests/test_yang_model.py:77]: Everytime data is reassigned, the memory referenced by it is going out of scope and should be freed.

File: src/sonic-yang-models/tests/yang_model_tests/tests/acl.json
Comment (joyas-joseph) [reply src/sonic-yang-models/tests/yang_model_tests/tests/acl.json:4]: I tried. But, couldn't find a way to do it from json files.

File: src/sonic-yang-models/tests/yang_model_tests/tests/acl.json
Comment (joyas-joseph) [reply src/sonic-yang-models/tests/yang_model_tests/tests/acl.json]: Removed eStr from the json file in this case and treated absence as []

**Review by anand-kumar-subramanian — DISMISSED** [review]

File: src/sonic-yang-models/tests/yang_model_tests/tests/acl.json
Comment (joyas-joseph) [reply src/sonic-yang-models/tests/yang_model_tests/tests/acl.json]: I have added "eStrKey" which will be used to index into defaultYANGFailure to get list of strings to match. "eStr" would be appended to this to get the final list.

File: src/sonic-yang-models/tests/yang_model_tests/tests/crm.json
Comment (joyas-joseph) [reply src/sonic-yang-models/tests/yang_model_tests/tests/crm.json:2]: Total time for the test config load
```
INFO:YANG-TEST:Time to load config:0:00:00.002771
```

**Comment by joyas-joseph** [comment]
retest this please

File: src/sonic-yang-models/tests/yang_model_tests/tests/port.json
Comment (joyas-joseph) [reply src/sonic-yang-models/tests/yang_model_tests/tests/port.json]: Updated the PR with the latest.

**Review by anand-kumar-subramanian — DISMISSED** [review]

File: src/sonic-yang-models/tests/yang_model_tests/test_yang_model.py
Comment (praveen-li) [reply src/sonic-yang-models/tests/yang_model_tests/test_yang_model.py:86]: To keep a Note for this: There are multiple references about json.load being slow, and with many long files here, it will be a concern sooner or later. So we would like to improve it in the future!!!!.  We are not fixing this concern in the current PR.

**Comment by praveen-li** [comment]
@joyas-joseph:
Add 'Signed-off-by: Praveen Chaudhary pchaudhary@linkedin.com' as well in this PR. 

So that questions about original content can be redirected to me in the future using GITHUB blame.

**Comment by joyas-joseph** [comment]
> @joyas-joseph:
> Add 'Signed-off-by: Praveen Chaudhary [pchaudhary@linkedin.com](mailto:pchaudhary@linkedin.com)' as well in this PR.
> Also, add a reference to the previous PR in the commit log, I think: #3730
> 
> So that questions about original content can be redirected to me in the future using GITHUB blame.

Update the PR.

**Comment by praveen-li** [comment]
@joyas-joseph thanks

@lguohan kindly merge this pr today if possible so that other prs can be on top of it. Thx a lot.

**Review by praveen-li — DISMISSED** [review]

**Comment by lguohan** [comment]
can you help to resolve the conflict?

**Review by anand-kumar-subramanian — APPROVED** [review]

**Review by lguohan — DISMISSED** [review]

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/buffers_defaults_t0.j2
Comment (praveen-li) [inline device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/buffers_defaults_t0.j2]: Hi Joyesh, We need rebase here. Seems something got messed up here, these files should not be in this diff. I remember I approve this PR, but kindly fix this last comment and I'll approve  again, thanks a lot

**Review by praveen-li — CHANGES_REQUESTED** [review]

**Review by praveen-li — DISMISSED** [review]

**Comment by praveen-li** [comment]
@lguohan kindly merge this pr before other yang PRs if possible, so that new prs can be on top of it. Thx a lot.

**Comment by lguohan** [comment]
@joyas-joseph , more yang related pr has been merged, can you resolve the conflict?

**Review by anand-kumar-subramanian — DISMISSED** [review]

**Comment by bhavini-gada** [comment]
@lguohan - Can you please merge this before merging any new PRs ?

**Review by praveen-li — APPROVED** [review]

**Review by lguohan — APPROVED** [review]

**Pull request merged**
